File: Tests/VIPSImageTransformTests.m
//
//  VIPSImageTransformTests.m
//  VIPSKitTests
//
//  Tests for VIPSImage+Transform: crop, rotate, flip, smart crop methods
//

#import "VIPSImageTestCase.h"

@interface VIPSImageTransformTests : VIPSImageTestCase
@end

@implementation VIPSImageTransformTests

#pragma mark - Crop Tests

- (void)testCrop {
    VIPSImage *image = [self createTestImageWithWidth:200 height:200];

    NSError *error = nil;
    VIPSImage *cropped = [image cropWithX:50 y:50 width:100 height:100 error:&error];

    XCTAssertNotNil(cropped, @"Should crop image");
    XCTAssertNil(error, @"Should not have error: %@", error);
    XCTAssertEqual(cropped.width, 100, @"Cropped width should match");
    XCTAssertEqual(cropped.height, 100, @"Cropped height should match");
}

- (void)testCropCorner {
    VIPSImage *image = [self createTestImageWithWidth:200 height:200];

    NSError *error = nil;
    VIPSImage *cropped = [image cropWithX:0 y:0 width:50 height:50 error:&error];

    XCTAssertNotNil(cropped, @"Should crop top-left corner");
    XCTAssertEqual(cropped.width, 50, @"Width should match");
    XCTAssertEqual(cropped.height, 50, @"Height should match");
}

- (void)testCropFullWidth {
    VIPSImage *image = [self createTestImageWithWidth:200 height:200];

    NSError *error = nil;
    VIPSImage *cropped = [image cropWithX:0 y:50 width:200 height:100 error:&error];

    XCTAssertNotNil(cropped, @"Should crop full width strip");
    XCTAssertEqual(cropped.width, 200, @"Width should be full");
    XCTAssertEqual(cropped.height, 100, @"Height should match");
}

#pragma mark - Rotation Tests

- (void)testRotate90 {
    VIPSImage *image = [self createTestImageWithWidth:200 height:100];

    NSError *error = nil;
    VIPSImage *rotated = [image rotateByDegrees:90 error:&error];

    XCTAssertNotNil(rotated, @"Should rotate image");
    XCTAssertNil(error, @"Should not have error: %@", error);
    XCTAssertEqual(rotated.width, 100, @"Width and height should swap");
    XCTAssertEqual(rotated.height, 200, @"Width and height should swap");
}

- (void)testRotate180 {
    VIPSImage *image = [self createTestImageWithWidth:200 height:100];

    NSError *error = nil;
    VIPSImage *rotated = [image rotateByDegrees:180 error:&error];

    XCTAssertNotNil(rotated, @"Should rotate image");
    XCTAssertEqual(rotated.width, 200, @"Width should stay same");
    XCTAssertEqual(rotated.height, 100, @"Height should stay same");
}

- (void)testRotate270 {
    VIPSImage *image = [self createTestImageWithWidth:200 height:100];

    NSError *error = nil;
    VIPSImage *rotated = [image rotateByDegrees:270 error:&error];

    XCTAssertNotNil(rotated, @"Should rotate image");
    XCTAssertEqual(rotated.width, 100, @"Width and height should swap");
    XCTAssertEqual(rotated.height, 200, @"Width and height should swap");
}

- (void)testRotateNegative90 {
    VIPSImage *image = [self createTestImageWithWidth:200 height:100];

    NSError *error = nil;
    VIPSImage *rotated = [image rotateByDegrees:-90 error:&error];

    XCTAssertNotNil(rotated, @"Should rotate image");
    XCTAssertEqual(rotated.width, 100, @"Width and height should swap");
    XCTAssertEqual(rotated.height, 200, @"Width and height should swap");
}

- (void)testRotate0 {
    VIPSImage *image = [self createTestImageWithWidth:200 height:100];

    NSError *error = nil;
    VIPSImage *rotated = [image rotateByDegrees:0 error:&error];

    XCTAssertNotNil(rotated, @"Should handle 0 degree rotation");
    XCTAssertEqual(rotated.width, 200, @"Width should stay same");
    XCTAssertEqual(rotated.height, 100, @"Height should stay same");
}

#pragma mark - Flip Tests

- (void)testFlipHorizontal {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSError *error = nil;
    VIPSImage *flipped = [image flipHorizontalWithError:&error];

    XCTAssertNotNil(flipped, @"Should flip image");
    XCTAssertNil(error, @"Should not have error: %@", error);
    XCTAssertEqual(flipped.width, image.width, @"Dimensions should not change");
    XCTAssertEqual(flipped.height, image.height, @"Dimensions should not change");
}

- (void)testFlipVertical {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSError *error = nil;
    VIPSImage *flipped = [image flipVerticalWithError:&error];

    XCTAssertNotNil(flipped, @"Should flip image");
    XCTAssertNil(error, @"Should not have error: %@", error);
    XCTAssertEqual(flipped.width, image.width, @"Dimensions should not change");
    XCTAssertEqual(flipped.height, image.height, @"Dimensions should not change");
}

#pragma mark - Auto Rotate Tests

- (void)testAutoRotate {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSError *error = nil;
    VIPSImage *rotated = [image autoRotateWithError:&error];

    XCTAssertNotNil(rotated, @"Should auto-rotate image");
    XCTAssertNil(error, @"Should not have error: %@", error);
    // Without EXIF orientation, dimensions should stay the same
}

#pragma mark - Smart Crop Tests

- (void)testSmartCrop {
    VIPSImage *image = [self createTestImageWithWidth:400 height:300];

    NSError *error = nil;
    VIPSImage *cropped = [image smartCropToWidth:200 height:200 interesting:VIPSInterestingAttention error:&error];

    XCTAssertNotNil(cropped, @"Should smart crop image");
    XCTAssertNil(error, @"Should not have error: %@", error);
    XCTAssertEqual(cropped.width, 200, @"Should match target width");
    XCTAssertEqual(cropped.height, 200, @"Should match target height");
}

- (void)testSmartCropStrategies {
    VIPSImage *image = [self createTestImageWithWidth:400 height:300];
    NSError *error = nil;

    // Test different strategies
    VIPSInteresting strategies[] = {
        VIPSInterestingNone,
        VIPSInterestingCentre,
        VIPSInterestingEntropy,
        VIPSInterestingAttention,
        VIPSInterestingLow,
        VIPSInterestingHigh
    };

    for (int i = 0; i < 6; i++) {
        VIPSImage *cropped = [image smartCropToWidth:200 height:200 interesting:strategies[i] error:&error];
        XCTAssertNotNil(cropped, @"Should smart crop with strategy %d", i);
        XCTAssertEqual(cropped.width, 200, @"Width should match with strategy %d", i);
        XCTAssertEqual(cropped.height, 200, @"Height should match with strategy %d", i);
    }
}

- (void)testSmartCropPortrait {
    VIPSImage *image = [self createTestImageWithWidth:200 height:400];

    NSError *error = nil;
    VIPSImage *cropped = [image smartCropToWidth:150 height:150 interesting:VIPSInterestingEntropy error:&error];

    XCTAssertNotNil(cropped, @"Should smart crop portrait image");
    XCTAssertEqual(cropped.width, 150, @"Should match target width");
    XCTAssertEqual(cropped.height, 150, @"Should match target height");
}

@end

File: Tests/VIPSImageTransformTests.swift
import XCTest
@testable import VIPSKit

final class VIPSImageTransformTests: VIPSImageTestCase {

    // MARK: - Crop

    func testCrop() throws {
        let image = createTestImage(width: 200, height: 200)
        let cropped = try image.crop(x: 50, y: 50, width: 100, height: 100)
        XCTAssertEqual(cropped.width, 100)
        XCTAssertEqual(cropped.height, 100)
    }

    func testCropCorner() throws {
        let image = createTestImage(width: 100, height: 100)
        let cropped = try image.crop(x: 0, y: 0, width: 50, height: 50)
        XCTAssertEqual(cropped.width, 50)
        XCTAssertEqual(cropped.height, 50)
    }

    func testCropFullWidth() throws {
        let image = createTestImage(width: 100, height: 200)
        let cropped = try image.crop(x: 0, y: 50, width: 100, height: 100)
        XCTAssertEqual(cropped.width, 100)
        XCTAssertEqual(cropped.height, 100)
    }

    // MARK: - Rotation

    func testRotate90() throws {
        let image = createTestImage(width: 200, height: 100)
        let rotated = try image.rotate(degrees: 90)
        XCTAssertEqual(rotated.width, 100)
        XCTAssertEqual(rotated.height, 200)
    }

    func testRotate180() throws {
        let image = createTestImage(width: 200, height: 100)
        let rotated = try image.rotate(degrees: 180)
        XCTAssertEqual(rotated.width, 200)
        XCTAssertEqual(rotated.height, 100)
    }

    func testRotate270() throws {
        let image = createTestImage(width: 200, height: 100)
        let rotated = try image.rotate(degrees: 270)
        XCTAssertEqual(rotated.width, 100)
        XCTAssertEqual(rotated.height, 200)
    }

    func testRotateNegative90() throws {
        let image = createTestImage(width: 200, height: 100)
        let rotated = try image.rotate(degrees: -90)
        XCTAssertEqual(rotated.width, 100)
        XCTAssertEqual(rotated.height, 200)
    }

    func testRotate0() throws {
        let image = createTestImage(width: 200, height: 100)
        let rotated = try image.rotate(degrees: 0)
        XCTAssertEqual(rotated.width, 200)
        XCTAssertEqual(rotated.height, 100)
    }

    // MARK: - Flip

    func testFlippedHorizontally() throws {
        let image = createTestImage(width: 100, height: 100)
        let flipped = try image.flippedHorizontally()
        XCTAssertEqual(flipped.width, 100)
        XCTAssertEqual(flipped.height, 100)
    }

    func testFlippedVertically() throws {
        let image = createTestImage(width: 100, height: 100)
        let flipped = try image.flippedVertically()
        XCTAssertEqual(flipped.width, 100)
        XCTAssertEqual(flipped.height, 100)
    }

    // MARK: - Auto Rotate

    func testAutoRotated() throws {
        let image = createTestImage(width: 100, height: 100)
        let rotated = try image.autoRotated()
        XCTAssertEqual(rotated.width, 100)
        XCTAssertEqual(rotated.height, 100)
    }

    func testAutoRotatedWithOrientation6() throws {
        // Orientation 6 = 90° CW rotation needed
        guard let path = pathForTestResource("rotated-6.jpg") else {
            XCTFail("Test resource rotated-6.jpg not found")
            return
        }
        let image = try VIPSImage(contentsOfFile: path)

        // Check that orientation metadata is present
        XCTAssertEqual(image.orientation, 6)

        // Original dimensions (before auto-rotate)
        XCTAssertEqual(image.width, 256)
        XCTAssertEqual(image.height, 256)

        // After auto-rotate, orientation should be applied
        let rotated = try image.autoRotated()
        // For a square image, dimensions stay the same
        XCTAssertEqual(rotated.width, 256)
        XCTAssertEqual(rotated.height, 256)
        // Orientation should be reset to 1 (or removed)
        XCTAssertTrue(rotated.orientation == nil || rotated.orientation == 1)
    }

    func testAutoRotatedWithOrientation3() throws {
        // Orientation 3 = 180° rotation needed
        guard let path = pathForTestResource("rotated-3.jpg") else {
            XCTFail("Test resource rotated-3.jpg not found")
            return
        }
        let image = try VIPSImage(contentsOfFile: path)

        // Check that orientation metadata is present
        XCTAssertEqual(image.orientation, 3)

        let rotated = try image.autoRotated()
        XCTAssertEqual(rotated.width, 256)
        XCTAssertEqual(rotated.height, 256)
        // Orientation should be reset
        XCTAssertTrue(rotated.orientation == nil || rotated.orientation == 1)
    }

    func testAutoRotatedPixelVerification() throws {
        // Load the rotated image and verify pixels are actually rotated
        guard let path = pathForTestResource("rotated-6.jpg") else {
            XCTFail("Test resource rotated-6.jpg not found")
            return
        }
        let image = try VIPSImage(contentsOfFile: path)
        let rotated = try image.autoRotated()

        // Sample a corner pixel to verify rotation happened
        // The asymmetric pattern has red in top-left (after rotation),
        // so we can verify the rotation was applied correctly
        let cornerColor = try rotated.pixelValues(atX: 10, y: 10)
        // After 90° CW rotation, the colors should have shifted
        XCTAssertNotNil(cornerColor)
    }

    // MARK: - Smart Crop

    func testSmartCropCentre() throws {
        let image = createTestImage(width: 200, height: 200)
        let cropped = try image.smartCrop(toWidth: 100, height: 100, interesting: .centre)
        XCTAssertEqual(cropped.width, 100)
        XCTAssertEqual(cropped.height, 100)
    }

    func testSmartCropEntropy() throws {
        let image = createTestImage(width: 200, height: 200)
        let cropped = try image.smartCrop(toWidth: 100, height: 100, interesting: .entropy)
        XCTAssertEqual(cropped.width, 100)
        XCTAssertEqual(cropped.height, 100)
    }

    func testSmartCropAttention() throws {
        let image = createTestImage(width: 200, height: 200)
        let cropped = try image.smartCrop(toWidth: 100, height: 100, interesting: .attention)
        XCTAssertEqual(cropped.width, 100)
        XCTAssertEqual(cropped.height, 100)
    }

    // MARK: - Async

    func testAsyncCropped() async throws {
        let image = createTestImage(width: 200, height: 200)
        let cropped = try await image.cropped(x: 50, y: 50, width: 100, height: 100)
        XCTAssertEqual(cropped.width, 100)
        XCTAssertEqual(cropped.height, 100)
    }

    func testAsyncCroppedCGRect() async throws {
        let image = createTestImage(width: 200, height: 200)
        let cropped = try await image.cropped(CGRect(x: 0, y: 0, width: 50, height: 50))
        XCTAssertEqual(cropped.width, 50)
        XCTAssertEqual(cropped.height, 50)
    }

    func testAsyncSmartCropped() async throws {
        let image = createTestImage(width: 200, height: 200)
        let cropped = try await image.smartCropped(toWidth: 100, height: 100, interesting: .centre)
        XCTAssertEqual(cropped.width, 100)
        XCTAssertEqual(cropped.height, 100)
    }

    func testAsyncSmartCroppedCGSize() async throws {
        let image = createTestImage(width: 200, height: 200)
        let cropped = try await image.smartCropped(to: CGSize(width: 100, height: 100), interesting: .centre)
        XCTAssertEqual(cropped.width, 100)
        XCTAssertEqual(cropped.height, 100)
    }
}
